File: RB130_exercises/medium_1/bubblesort.rb
# Bubble sort with blocks

# Bubble Sort Algorithm:
# - Multiple iterations through the array. On each pass, each pair of 
#   consecutive elements is compared. 
# - If the first element is greater than the second, the two elements
#   are swapped; otherwise, they remain in place.
# - If a complete iteration is made without any swaps, the sorting is 
#   complete.

# Write a bubble sort method that takes a block specifying how to sort
# the array.

# Input: 
# - An array (to be sorted, destructively)
# - A block, specifying how elements should be sorted (optional)
#   - If no block is given, default to sorting in ascending order.

# Output: 
# - The same array, sorted

# Examples:
# array = [5, 3, 7]
# bubble_sort!(array) # Default to sorting in ascending order if no block
# array == [3, 5, 7]

# array = [5, 3, 7]
# bubble_sort!(array) { |first, second| first >= second }
# # If the block returns true(thy?), do not swap the two elements.
# array == [7, 5, 3]

# Data/Algorithm:
# - Array given as input
# - Loop through the array. For each loop:
#   - Set a flag, swapped = false
#   - Iterate through the array, excluding the last element. For each element:
#     - Compare the current element against the next element.
#       - If a block is given, yield current and next elements. Otherwise, compare
#         current and next elements using size.
#       - If the block returns falsey, swap 2 elements.
#       - If the comparison returns true, swap 2 elements.
#     - If a swap is made, set swapped = true
#   - EOI => Go back to loop
# - Break loop if swapped is still false.
# => Return array.

# def bubble_sort!(array)
#   loop do
#     swap_made = false

#     array[0..-2].each_with_index do |element, index|
#       next_element = array[index + 1]
#       skip = block_given? ? yield(element, next_element) : element <= next_element
#       next if skip 

#       array[index], array[index + 1] = array[index + 1], array[index]
#       swap_made = true
#     end
#     break unless swap_made
#   end
#   array
# end

# array = [5, 3]
# bubble_sort!(array)
# p array == [3, 5]

# array = [5, 3, 7]
# bubble_sort!(array) { |first, second| first >= second }
# p array == [7, 5, 3]

# array = [6, 2, 7, 1, 4]
# bubble_sort!(array)
# p array == [1, 2, 4, 6, 7]

# array = [6, 12, 27, 22, 14]
# bubble_sort!(array) { |first, second| (first % 7) <= (second % 7) }
# p array == [14, 22, 12, 6, 27]

# array = %w(sue Pete alice Tyler rachel Kim bonnie)
# bubble_sort!(array)
# p array == %w(Kim Pete Tyler alice bonnie rachel sue)

# array = %w(sue Pete alice Tyler rachel Kim bonnie)
# bubble_sort!(array) { |first, second| first.downcase <= second.downcase }
# p array == %w(alice bonnie Kim Pete rachel sue Tyler)


# Further Exploration ========================
# Modify your solution so that it only passes one element to the block at a time.
# - The block should perform a transformation on the argument, and bubblesort!
#   itself should use <= to compare the two values.

# Algorithm:
# - loop through the array; iterate through the array. for each element:
#   - Retrieve next element.
#   - If a block is given:
#     - yield the current and next element to the block (separately), then 
#       compare the block return values using <=.
#     - If the current element is less than/equal to next element, skip to next
#       iteration
#   - If no block is given:
#     - compare current and next elements directly using <= 


# Example:
# def bubble_sort_by!(array)
#   loop do
#     swap_made = false

#     array[0..-2].each_with_index do |element, index|
#       next_element = array[index + 1]
      
#       skip = if block_given?
#                yield(element) <= yield(next_element)
#              else
#                element <= next_element
#              end       
      
#       next if skip

#       array[index], array[index + 1] = array[index + 1], array[index]
#       swap_made = true
#     end
#     break unless swap_made
#   end
#   array
# end

# array = %w(sue Pete alice Tyler rachel Kim bonnie)
# bubble_sort_by!(array) { |value| value.downcase }
# # Sort the array using the values obtained by calling downcase on each 
# #  element of the array. (like sort_by)
# p array == %w(alice bonnie Kim Pete rachel sue Tyler)

File: javaScriptExercises/medium/bubbleSort.js
// Bubble Sort
// - The bubble sort works by making multiple passes through an array.
// - On each pass, consecutive element pairs are compared. If the 1st
//   value is greater than the 2nd, the two elements are swapped; 
//   otherwise, they remain in place. 
// - This process repeats until a complete pass is made without 
//   performing any swaps, at which point the array is considered 
//   fully sorted.

// Questions:
// 1) Can the array ever be missing or non-array?
// => No
// 2) Can the array be empty or only contain one element?
// => Yes => Return array as-is
// 3) Can the array contain non-primitive values?
// => No
// 4) What should the function return?
// => undefined


// Example:
// [6, 2, 7, 1, 4]
// Pass 1:
//  [6, 2]: 6 > 2 => Swap    => [2, 6, 7, 1, 4]
//  [6, 7]: 6 < 7 => No Swap => [2, 6, 7, 1, 4]
//  [7, 1]: 7 > 1 => Swap    => [2, 6, 1, 7, 4]
//  [7, 4]: 7 > 4 => Swap    => [2, 6, 1, 4, 7]

// Pass 2:
//  [2, 6]: 2 < 6 => No Swap => [2, 6, 1, 4, 7]
//  [6, 1]: 6 > 1 => Swap    => [2, 1, 6, 4, 7]
//  [6, 4]: 6 > 4 => Swap    => [2, 1, 4, 6, 7]
// * Each pass of a bubble sort will fully sort the nth largest element
//   - So each subsequent pass can look at one fewer element pair 

// Pass 3:
//  [2, 1]: 2 > 1 => Swap    => [1, 2, 4, 6, 7]
//  [2, 4]: 2 < 4 => No Swap => [1, 2, 4, 6, 7]

// Pass 4: 
//  All elements are fully sorted (no swaps will occur)


// Write a function that takes an array as an argument and sorts
// it in-place (mutating) using the bubble sort algorithm.

// Rules / Patterns:
// For any array of length n:
// - We will make n - 1 passes through the array before it is fully 
//   sorted.
//   > for (let pass = 1; pass < array.length)

// - On the first pass, the number of iterations is identical to the
//   number of total passes required
// - On each subsequent pass, the number of iterations decreases by 1
//   on each iteration, because each pass will fully sort the 
//   p-th largest element.
// > In other words, each pass will involve array.length - passNumber
//   # of iterations

// - On each pass, we iterate (array.length - passNumber) times;
//   > for (let index = 0; index < array.length - passNumber)

// - On each iteration, we retrieve the element at the current index
//   as well as the next index, and compare them: 
//   - If a > b, swap the elements in-place
//   - Otherwise, do nothing (continue)

// Data: 
// > Input: Array of elements
// - Keep as array
// - Make array.length - 1 passes through the array
// - On each pass, iterate x times, where x = array.length - passNumber
//   eg. For an array of 5 elements, on pass 1, we iterate 4 times.
//       On Pass 2, we iterate 3 times; etc...


// Algorithm:
// Given an array of elements as input:
// 1) [GUARD] - If the array length is less than 2, return array as-is
// 2) Initialize a loop for the number of passes through the array, 
//    which can be calculated as 1 up to the array's length - 1. 
// For each pass:
// 3) Initialize a loop for iterating through -some of- the array 
//    elements; the indexes of interest can be calculated as index 0
//    to index array.length - current pass number.
// For each iteration:
// 4) Compare the current element array[index] with the next element
//    array[index + 1]
// 5) If currentElement > nextElement, swap the two elements in-place:
//    [ array[index], array[index + 1] = [ array[index + 1], array[index] ]
// At the end of all passes, the array should be fully sorted.
// Return undefined.

// function bubbleSort(array) {
//   const arrayLength = array.length;
//   if (arrayLength < 2) return array;

//   // Passes
//   for (let passNumber = 1; passNumber < arrayLength; passNumber++) {
//     // Iterations
//     console.log(passNumber)
//     for (let index = 0; index < arrayLength - passNumber; index++) {
//       let [ currentElement, nextElement ] = [ array[index], array[index + 1] ];
      
//       if (currentElement > nextElement) {
//         [ array[index], array[index + 1]] = [ array[index + 1], array[index] ];
//       }
//     }
//   }
// }

// Refactor: Implement a swapped check so that if a complete pass is
// made without any swaps, the pass-through iteration can end (because
// the array might be fully-sorted in less than the max number of
// passes.)

// How?
// - On each pass, set a variable (swapped = false) before iteration
//   begins
// - During iteration, if a swap is made, set swapped = true
// - at the end of each pass, if swapped is still false, exit pass
//   iteration

function bubbleSort(array) {
  const arrayLength = array.length;
  if (arrayLength < 2) return array;

  // Passes
  for (let passNumber = 1; passNumber < arrayLength; passNumber++) {
    let noSwapsMade = true;
    // Iterations
    for (let index = 0; index < arrayLength - passNumber; index++) {
      let [ currentElement, nextElement ] = [ array[index], array[index + 1] ];
      
      if (currentElement > nextElement) {
        [ array[index], array[index + 1]] = [ array[index + 1], array[index] ];
        noSwapsMade = false;
      }
    }

    if (noSwapsMade) break;
  }
}


// Examples:
let arrayLength5 = [6, 2, 7, 1, 4];
bubbleSort(arrayLength5);
console.log(arrayLength5); //=> [1, 2, 4, 6, 7];

let arrayLength2 = [2, 1];
bubbleSort(arrayLength2);
console.log(arrayLength2); //=> [1, 2]

let alreadySorted = [1, 2, 3, 4, 5];
bubbleSort(alreadySorted);
console.log(alreadySorted); //=> [1, 2, 3, 4, 5]


// String Sorting
let stringArray = ['Sue', 'Pete', 'Alice', 'Tyler', 'Rachel', 'Kim', 'Bonnie'];
bubbleSort(stringArray);
console.log(stringArray); // ["Alice", "Bonnie", "Kim", "Pete", "Rachel", "Sue", "Tyler"]

// Edge Cases:
// - Empty Array
let emptyArray = [];
bubbleSort(emptyArray);
console.log(emptyArray); // []

// - One-element Array
let arrayLength1 = [1];
bubbleSort(arrayLength1); 
console.log(arrayLength1); // [1]
